### DISCUSSION TITLE
In test, does GraphSage needs both training and test time node?

### DISCUSSION THREAD

**Discussion by hosseinghafarian** [pyg-team/pytorch_geometric#7916] [Q&A]
It is emphasized in the graph sage paper that it is inductive. Moreover, the algorithm performs neighbor sampling which enables it to scale much better with respect to GCN. I have two inter-related questions. May be my understanding of graphsage is incorrect or incomplete. 

1- Can we batch train using neighbor sampling without having previous batches samples? My own answer is, yes, we can. One just needs to compute embeddings from possibly new samples and their possibly new neighbors. However, based on just the current batch. 
2- Then in test time, how to test for newly arrived samples? How to generate embeddings for these new samples? Are we required to store all training time samples?

**Comment by hosseinghafarian** [comment]
This colab [notebook](https://colab.research.google.com/drive/1Hmftn5d2IxeQNYHMs8C3aZDgNyR6WTRP#scrollTo=Vzy64TKsLAXA) seems to suggest that we must have all samples. Thank you for any hints or discussions.

**Comment by LukeLIN-web** [comment]
It is weird to use "sample" to refer to node/vertice.
You could refer https://github.com/pyg-team/pytorch_geometric/blob/master/examples/reddit.py 
The graph has been split.
```
train_loader = NeighborLoader(data, input_nodes=data.train_mask,
                              num_neighbors=[25, 10], shuffle=True, **kwargs)
```
It will iterate all nodes in the train set. 
For example, batch size = 1024, the sampler will sample  <= 1024*25*10 as a subgraph. 

In test time 
```
subgraph_loader = NeighborLoader(copy.copy(data), input_nodes=None,
                                 num_neighbors=[-1], shuffle=False, **kwargs)
``` 

It will iterate all nodes, evaluate the test set, and train set separately. You could also evaluate the test set only.
